### ISSUE TITLE
Cannot install an NPM package: npm'' finished with non-zero exit value 139

### ISSUE THREAD

**Issue by loxal** [node-gradle/gradle-node-plugin#46]
* What went wrong:
Execution failed for task ':installJest'.
> Process 'command '/home/alex/Desktop/kotlin-examples/gradle/js-tests/jest/.gradle/nodejs/node-v10.14.0-linux-x64/bin/npm'' finished with non-zero exit value 139

I am using:
* Gradle 5.6.4
* openjdk 11.0.5-ea 2019-10-15

> task installJest(type: NpmTask) {
>     args = ['install', 'jest']
> }

**Comment by bsautel** [comment]
I tried to reproduce this issue but I did not succeed.

Do you have any extra information in the build output? It sounds like npm fails to install this package, in this case, it should probably print something that Gradle should print to the build output.

**Comment by deepy** [comment]
exit value 139 indicates a segmentation fault, so this might not be related to the plugin.

could you test if the same operation works manually in node/npm?

**Comment by loxal** [comment]
I have tried to call the npm binaries directly and it worked. I do not have the environment anymore but same thing happened with com.moowork.gradle:gradle-node-plugin:1.3.0 & 1.3.1, however with com.moowork.gradle:gradle-node-plugin:1.2.0 & Gradle 5.6.4 everything worked find. This is also my current environment I am using for a PoC. However most likely I might use npm/node directly as the Gradle plugin wrapper solution seems to be too fragil. There were also issues when running it with Gradle 6.0 and/or Java 13. 

Sorry that I cannot provide more specific info. I already wasted too much time on this PoC not because of your specific plugin though :)

**Comment by loxal** [comment]
I see the "original" plugin is not under development anymore. So I try my best to help you, @deepy 

Running your plugin with Gradle 6.0 gives me:

> alex@sun:~/my/project/loxal/page-finder/js-jest-test$ ./gradlew test
> > Task :installJest FAILED
> 
> FAILURE: Build failed with an exception.
> 
> * What went wrong:
> Execution failed for task ':installJest'.
> > Process 'command '/home/alex/my/project/loxal/page-finder/js-jest-test/.gradle/nodejs/node-v10.14.0-linux-x64/bin/npm'' finished with non-zero exit value 139
> 
> * Try:
> Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
> 
> * Get more help at https://help.gradle.org
> 
> Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
> Use '--warning-mode all' to show the individual deprecation warnings.
> See https://docs.gradle.org/6.0/userguide/command_line_interface.html#sec:command_line_warnings
> 
> BUILD FAILED in 4s
> 4 actionable tasks: 4 executed
> alex@sun:~/my/project/loxal/page-finder/js-jest-test$ ./gradlew test
> > Task :installJest FAILED
> 
> FAILURE: Build failed with an exception.
> 
> * What went wrong:
> Execution failed for task ':installJest'.
> > Process 'command '/home/alex/my/project/loxal/page-finder/js-jest-test/.gradle/nodejs/node-v10.14.0-linux-x64/bin/npm'' finished with non-zero exit value 139
> 
> * Try:
> Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
> 
> * Get more help at https://help.gradle.org
> 
> Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
> Use '--warning-mode all' to show the individual deprecation warnings.
> See https://docs.gradle.org/6.0/userguide/command_line_interface.html#sec:command_line_warnings
> 
> BUILD FAILED in 1s
> 4 actionable tasks: 1 executed, 3 up-to-date
> alex@sun:~/my/project/loxal/page-finder/js-jest-test$ /home/alex/my/project/loxal/page-finder/js-jest-test/.gradle/nodejs/node-v10.14.0-linux-x64/bin/npm version
> { npm: '6.4.1',
>   ares: '1.15.0',
>   cldr: '34.0',
>   http_parser: '2.8.0',
>   icu: '63.2',
>   modules: '64',
>   napi: '3',
>   nghttp2: '1.37.0',
>   node: '10.15.2',
>   openssl: '1.1.1c',
>   tz: '2019a',
>   unicode: '11.0',
>   uv: '1.30.1',
>   v8: '6.8.275.32-node.12',
>   zlib: '1.2.11' }
> alex@sun:~/my/project/loxal/page-finder/js-jest-test$ 


As you see the binary seems to be ok.

**Comment by loxal** [comment]
With Gradle 5.6.4 and openjdk version "11.0.5-ea" 2019-10-15


> * What went wrong:
> Execution failed for task ':installJest'.
> > Process 'command '/home/alex/my/project/loxal/page-finder/js-jest-test/.gradle/nodejs/node-v10.14.0-linux-x64/bin/npm'' finished with non-zero exit value 139
> 
> * Try:
> Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
> 
> * Get more help at https://help.gradle.org
> 
> BUILD FAILED in 12s
> 4 actionable tasks: 4 executed
>

**Comment by bsautel** [comment]
Thanks for all that.

1. Your build use some deprecated Gradle 6 features that will be removed in Gradle 7. As far as I know, if you use the latest version of this plugin (2.2.0), it should not come from it. But to be sure, could you run your build that prints this warning with the `--warning-mode all` option? Note that this should not be related to your install task that does not work.
2. What happens if you manually run the `npm install jest` using the same npm binary as the one you used to ask the version?
3. According to your logs, I can see that you ask the plugin to download and install Node.js. Your `npm --version` shows something strange. The Node.js version seems to be 10.14.0 (in the directory name, that's the default version), but npm runs in a 10.15.2 Node.js. 🤔 I just come to make some tests in local and I can reproduce that. In fact, the `.gradle/nodejs/node-v10.14.0-linux-x64/bin/npm` file has `#!/usr/bin/env node` as shebang, so it uses the globally installed Node.js instead of the one installed by the plugin. You probably have Node.js 10.15.2 globally installed on your system, haven't you? This is not the expected behavior and is probably due to a bug in the plugin.

**Comment by loxal** [comment]
Ok, I've spent a bit more time... the problem turned out to be 

> task installJest(type: NpmTask) {
>     setNpmCommand("install", "jest")
> }

this did not work with the default Node version. So after doing _installJest_ with the original v1.2.0 plugin I could execute this plugins v2.2.0. But I tried further and after 

> node {
>     version = "12.13.0"
>     download = true
> }

...setting the Node version to the latest LTS release, everything, incl. installJest went smoothly.

**Comment by deepy** [comment]
> In fact, the `.gradle/nodejs/node-v10.14.0-linux-x64/bin/npm` file has `#!/usr/bin/env node` as shebang

Which should be fine as we do the Path song and dance at https://github.com/node-gradle/gradle-node-plugin/blob/master/src/main/groovy/com/moowork/gradle/node/exec/ExecRunner.groovy#L88-L98

> so it uses the globally installed Node.js instead of the one installed by the plugin.

With that said, I can't say that I'm certain that this always works.

>Ok, I've spent a bit more time... the problem turned out to be

That's a huge relief :-)

**Comment by bsautel** [comment]
Nice! I did not think about the _path song and dance_ and you're right, it the path is right, the shebang should lead to use the right Node.js interpreter, which is not the case when we launch directly the binary from the terminal without setting the path.
I discovered that by running it manually but not through the plugin, so my tests are not valid.
I will be off for three weeks as of tomorrow, but when I am back I will try to add some tests to ensure we use the right interpreter. As you say, it should at least work in most cases.

We still don't really know why the task works with the latest Node.js version and not with an older one. But as this issue is resolved, maybe we should close it?
